File: backend/src/main/java/com/againspring/service/ai/LlmErrorSignatures.java
package com.againspring.service.ai;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;

/**
 * LLM 오류·거절·누출 시그니처 SSOT 로더.
 * 데이터는 docs/shared/policies/llm-error-signatures.json 한 파일이다(컨테이너 /app/shared/docs/policies에 :ro 마운트).
 * 파일이 없으면 기동 실패 — 시그니처 없이 게시하는 것이 더 위험하다(절대 규칙 #7).
 */
public final class LlmErrorSignatures {
    public static final String ENV_PATH = "LLM_ERROR_SIGNATURES_PATH";
    private static final List<String> CANDIDATES = List.of(
            "/app/shared/docs/policies/llm-error-signatures.json",
            "../docs/shared/policies/llm-error-signatures.json");      // backend

    private static volatile LlmErrorSignatures instance;

    private final List<String> signatures;
    private final List<Pattern> promptLeakPatterns;
    private final double koreanRatioMin;
    private final int koreanCheckMinChars;

    private LlmErrorSignatures(List<String> signatures, List<Pattern> leaks, double ratio, int minChars) {
        this.signatures = List.copyOf(signatures);
        this.promptLeakPatterns = List.copyOf(leaks);
        this.koreanRatioMin = ratio;
        this.koreanCheckMinChars = minChars;
    }

    public static LlmErrorSignatures get() {
        LlmErrorSignatures local = instance;
        if (local == null) {
            synchronized (LlmErrorSignatures.class) {
                local = instance;
                if (local == null) instance = local = load();
            }
        }
        return local;
    }

    static LlmErrorSignatures load() {
        Path path = resolvePath();
        try {
            JsonNode root = new ObjectMapper().readTree(Files.readString(path));
            List<String> sigs = new ArrayList<>();
            for (JsonNode n : root.path("signatures")) {
                String s = n.asText("").trim().toLowerCase(Locale.ROOT);
                if (!s.isEmpty()) sigs.add(s);
            }
            List<Pattern> leaks = new ArrayList<>();
            for (JsonNode n : root.path("prompt_leak_patterns")) leaks.add(Pattern.compile(n.asText()));
            if (sigs.isEmpty()) throw new IllegalStateException("llm-error-signatures.json has no signatures: " + path);
            return new LlmErrorSignatures(sigs, leaks,
                    root.path("korean_ratio_min").asDouble(0.10),
                    root.path("korean_check_min_chars").asInt(20));
        } catch (IOException e) {
            throw new IllegalStateException("cannot read llm-error-signatures.json at " + path, e);
        }
    }

    private static Path resolvePath() {
        String env = System.getenv(ENV_PATH);
        if (env != null && !env.isBlank() && Files.isRegularFile(Path.of(env))) return Path.of(env);
        for (String c : CANDIDATES) {
            Path p = Path.of(c);
            if (Files.isRegularFile(p)) return p.toAbsolutePath().normalize();
        }
        throw new IllegalStateException("llm-error-signatures.json not found; set " + ENV_PATH
                + " or mount /app/shared/docs/policies/llm-error-signatures.json");
    }

    public List<String> signatures() { return signatures; }
    public List<Pattern> promptLeakPatterns() { return promptLeakPatterns; }
    public double koreanRatioMin() { return koreanRatioMin; }
    public int koreanCheckMinChars() { return koreanCheckMinChars; }

    /** {@code lower}는 호출자가 소문자로 넘긴다. */
    public boolean containsSignature(String lower) {
        for (String s : signatures) if (lower.contains(s)) return true;
        return false;
    }

    /** 한국어 AI 콘텐츠에 한글이 사실상 없으면(비율 < koreanRatioMin) 영어 거절·오류로 판정. */
    public boolean hasInsufficientKorean(String text) {
        long significant = text.chars().filter(c -> c > 32).count();
        if (significant < koreanCheckMinChars) return false;
        long korean = text.chars().filter(c ->
                (c >= 0xAC00 && c <= 0xD7A3) || (c >= 0x1100 && c <= 0x11FF) || (c >= 0x3130 && c <= 0x318F)).count();
        return (double) korean / significant < koreanRatioMin;
    }

    public boolean hasPromptLeak(String text) {
        for (Pattern p : promptLeakPatterns) if (p.matcher(text).find()) return true;
        return false;
    }
}

File: ai-user/learning/app/services/llm_error_signatures.py
"""LLM 오류·거절·누출 시그니처 SSOT 로더 (docs/shared/policies/llm-error-signatures.json)."""
from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path

ENV_PATH = "LLM_ERROR_SIGNATURES_PATH"
_CANDIDATES = (
    Path("/app/shared/docs/policies/llm-error-signatures.json"),
    Path(__file__).resolve().parents[4] / "docs" / "shared" / "policies" / "llm-error-signatures.json",
)


@dataclass(frozen=True)
class Signatures:
    signatures: tuple[str, ...]
    prompt_leak_patterns: tuple[re.Pattern, ...]
    korean_ratio_min: float
    korean_check_min_chars: int

    def contains_signature(self, lower: str) -> bool:
        return any(sig in lower for sig in self.signatures)

    def has_insufficient_korean(self, text: str) -> bool:
        significant = sum(1 for ch in text if ord(ch) > 32)
        if significant < self.korean_check_min_chars:
            return False
        korean = sum(1 for ch in text
                     if "가" <= ch <= "힣" or "ᄀ" <= ch <= "ᇿ" or "㄰" <= ch <= "㆏")
        return korean / significant < self.korean_ratio_min

    def has_prompt_leak(self, text: str) -> bool:
        return any(p.search(text) for p in self.prompt_leak_patterns)


def _resolve() -> Path:
    env = os.getenv(ENV_PATH)
    if env and Path(env).is_file():
        return Path(env)
    for c in _CANDIDATES:
        if c.is_file():
            return c
    raise RuntimeError(f"llm-error-signatures.json not found; set {ENV_PATH}")


@lru_cache(maxsize=1)
def load() -> Signatures:
    data = json.loads(_resolve().read_text(encoding="utf-8"))
    sigs = tuple(s.strip().lower() for s in data["signatures"] if s.strip())
    if not sigs:
        raise RuntimeError("llm-error-signatures.json has no signatures")
    # Java (?m) 플래그 → Python re.M
    leaks = tuple(re.compile(p.replace("(?m)", ""), re.M) for p in data.get("prompt_leak_patterns", []))
    return Signatures(sigs, leaks, float(data.get("korean_ratio_min", 0.10)), int(data.get("korean_check_min_chars", 20)))


def looks_like_llm_error(text: str | None) -> bool:
    if not text or not text.strip():
        return False
    s = load()
    if s.has_insufficient_korean(text):
        return True
    lower = re.sub(r"\s+", " ", text).strip().lower()
    return s.contains_signature(lower)
